libdnf5/base/goal.cpp

```diff
@@ -509,6 +509,13 @@ GoalProblem Goal::Impl::add_specs_to_goal(base::Transaction & transaction) {
                         base, query, settings.advisory_filter.value(), cfg_main.get_obsoletes_option().get_value());
                 }
 
+                // Remove packages of versions lower than the installed ones.
+                // Substract downgrades instead of filter_upgrades() to keep
+                // installed packages in the query.
+                rpm::PackageQuery tmp_downgrades(query);
```

Comment:
I am quite curious, why you want to keep packages with the same version as installed in the query?

Reply:
Well, I don't :) I've just overlooked it. Thanks for noticing.

Reply:
Fixed.

Reply:
> I recall that "dnf upgrade" reinstalls packages which were originally installed unsigned from a local build and then they appeared properly signed in a repository. That could warrant leaving same versions in the query.

I need to check this. We do not have a test for it, I suppose?

Reply:
@ppisar it seems that you are refering to behaviour described in this libsolv issue: https://github.com/openSUSE/libsolv/issues/287
The trigger for re-installation was not whether the package was or was not signed, but more generally that the package was changed (for example has different build time).
This behavior was changed in dnf4 here: https://github.com/rpm-software-management/libdnf/pull/786 (and the same behavior we have in dnf5).
So dnf4 does not reinstall package with the same nevra but different buildtime for quite a long time. But distrosync command should trigger the reinstallation.

Reply:
Yes, that libsolv ticket describes my experience. If DNF stopped doing it, the then j-mracek's proposal is fine.